compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/IrSourceCompilerForInline.kt

```diff
@@ -59,19 +58,27 @@ class IrSourceCompilerForInline(
     override val lazySourceMapper: DefaultSourceMapper
         get() = codegen.classCodegen.getOrCreateSourceMapper()
 
-    override fun generateLambdaBody(adapter: MethodVisitor, jvmMethodSignature: JvmMethodSignature, lambdaInfo: ExpressionLambda): SMAP {
-        lambdaInfo as? IrExpressionLambdaImpl ?: error("Expecting ir lambda, but $lambdaInfo")
+    override fun generateLambdaBody(adapter: MethodVisitor, jvmMethodSignature: JvmMethodSignature, lambdaInfo: ExpressionLambda): SMAP =
+        throw AssertionError("unreachable; IrExpressionLambdaImpl overrides generateLambdaBody")
```

Comment:
This method still looks strange: base implementation is used from only one place (ExpressionLambda). So maybe switch adapter here to [Ir]ExpressionLambda and move adapter related logic from ExpressionLambda also to sourceCompilerForInline